File: sample/src/main/java/com/github/lany192/box/sample/mvp/main/MainActivity.java
package com.github.lany192.box.sample.mvp.main;

import android.os.Bundle;
import android.view.KeyEvent;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.github.lany192.box.activity.ActivityConfig;
import com.github.lany192.box.activity.BaseActivity;
import com.github.lany192.box.sample.R;
import com.github.lany192.box.sample.fragment.HelloFragment;
import com.github.lany192.box.sample.mvp.main.city.CityFragment;
import com.github.lany192.box.sample.mvp.main.index.IndexFragment;
import com.github.lany192.box.sample.fragment.MyFragment;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.gyf.immersionbar.ImmersionBar;
import com.hjq.toast.ToastUtils;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class MainActivity extends BaseActivity implements MainContract.View {
    ViewPager2 mViewPager2;
    BottomNavigationView mBottomNavigationView;
    @Inject
    MainPresenter mPresenter;

    // 第一次按退出的时间
    private long mLastClickTime = 0;

    @NonNull
    @Override
    public ActivityConfig getConfig() {
        return ActivityConfig.builder()
                .layoutId(R.layout.activity_main)
                .statusBarDarkFont(true)
                .transparentStatusBar(true)
                .build();
    }

    @Override
    protected void init(Bundle savedInstanceState) {
        mViewPager2 = findViewById(R.id.main_viewpager);
        mBottomNavigationView = findViewById(R.id.main_navigation_bar);
        mViewPager2.setUserInputEnabled(false);
        mViewPager2.setAdapter(new FragmentStateAdapter(this) {

            @Override
            public int getItemCount() {
                return 4;
            }

            @NonNull
            @Override
            public Fragment createFragment(int position) {
                switch (position) {
                    case 0:
                        return new IndexFragment();
                    case 1:
                        return new HelloFragment();
                    case 2:
                        return new CityFragment();
                    case 3:
                        return new MyFragment();
                    default:
                        return new IndexFragment();
                }
            }
        });
        mViewPager2.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                mBottomNavigationView.getMenu().getItem(position).setChecked(true);
            }
        });
        mBottomNavigationView.setOnNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.menu_main_index:
                    ImmersionBar.with(this).statusBarDarkFont(true).init();
                    mViewPager2.setCurrentItem(0, false);
                    return true;
                case R.id.menu_main_pic:
                    ImmersionBar.with(this).statusBarDarkFont(true).init();
                    mViewPager2.setCurrentItem(1, false);
                    return true;
                case R.id.menu_main_city:
                    ImmersionBar.with(this).statusBarDarkFont(true).init();
                    mViewPager2.setCurrentItem(2, false);
                    return true;
                case R.id.menu_main_my:
                    ImmersionBar.with(this).statusBarDarkFont(false).init();
                    mViewPager2.setCurrentItem(3, false);
                    return true;
                default:
            }
            return false;
        });
        mPresenter.hello();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if ((System.currentTimeMillis() - mLastClickTime) > 3000) {
                ToastUtils.show("再按一次退出" + getString(R.string.app_name));
                mLastClickTime = System.currentTimeMillis();
                return false;
            }
        }
        return super.onKeyDown(keyCode, event);
    }
}

File: sample/src/main/java/com/lany192/box/sample/ui/main/MainActivity.kt
package com.lany192.box.sample.ui.main

import android.os.Bundle
import android.view.KeyEvent
import android.view.MenuItem
import androidx.viewpager2.widget.ViewPager2
import com.alibaba.android.arouter.facade.annotation.Route
import com.github.lany192.arch.activity.ViewModelActivity
import com.github.lany192.arch.tab.TabAdapter
import com.github.lany192.arch.tab.TabItem
import com.github.lany192.log.LogUtils
import com.gyf.immersionbar.ImmersionBar
import com.hjq.toast.Toaster
import com.lany192.box.network.data.bean.UserInfo
import com.lany192.box.sample.R
import com.lany192.box.sample.databinding.ActivityMainBinding
import com.lany192.box.sample.ui.main.discover.DiscoverBuilder
import com.lany192.box.sample.ui.main.index.IndexBuilder
import com.lany192.box.sample.ui.main.menus.MenusBuilder
import com.lany192.box.sample.ui.main.message.MessageBuilder
import com.lany192.box.sample.ui.main.my.MyBuilder
import com.lany192.box.sample.ui.user.UserViewModel
import dagger.hilt.android.AndroidEntryPoint

@Route(path = "/ui/main")
@AndroidEntryPoint
class MainActivity : ViewModelActivity<MainViewModel, ActivityMainBinding>() {
    // 第一次按退出的时间
    private var mLastClickTime: Long = 0
    private lateinit var userViewModel: UserViewModel

    override fun getViewBinding(): ActivityMainBinding {
        return ActivityMainBinding.inflate(layoutInflater)
    }

    override fun initImmersionBar() {
        return ImmersionBar.with(this).init()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Toaster.show("构建时间：" + getString(R.string.apk_build_time))
//        ViewUtils.setGrayStyle(this, true)
        userViewModel = getAndroidViewModel(UserViewModel::class.java)
        userViewModel.userInfo.observe(this) { userInfo: UserInfo -> Toaster.show("首页：" + userInfo.name) }
        val items = mutableListOf<TabItem>()
        items.add(TabItem("测试", MenusBuilder.getFragment()))
        items.add(TabItem("首页", IndexBuilder.getFragment()))
        items.add(TabItem("发现", DiscoverBuilder.getFragment()))
        items.add(TabItem("消息", MessageBuilder.getFragment()))
        items.add(TabItem("我的", MyBuilder.getFragment()))
        binding.viewpager.isUserInputEnabled = false
        binding.viewpager.offscreenPageLimit = items.size
        binding.viewpager.adapter = TabAdapter(this, items)
        binding.viewpager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                binding.navigationView.menu.getItem(position).isChecked = true
            }
        })
        //去掉选中图标的背景色
        binding.navigationView.isItemActiveIndicatorEnabled = false
        binding.navigationView.setOnItemSelectedListener { item: MenuItem ->
            when (item.itemId) {
                R.id.menu_main_index -> {
                    binding.viewpager.setCurrentItem(0, false)
                    return@setOnItemSelectedListener true
                }

                R.id.menu_main_menus -> {
                    binding.viewpager.setCurrentItem(1, false)
                    return@setOnItemSelectedListener true
                }

                R.id.menu_main_pic -> {
                    binding.viewpager.setCurrentItem(2, false)
                    return@setOnItemSelectedListener true
                }

                R.id.menu_main_city -> {
                    binding.viewpager.setCurrentItem(3, false)
                    return@setOnItemSelectedListener true
                }

                R.id.menu_main_my -> {
                    binding.viewpager.setCurrentItem(4, false)
                    return@setOnItemSelectedListener true
                }

                else -> {}
            }
            false
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            if (System.currentTimeMillis() - mLastClickTime > 3000) {
                Toaster.show("再按一次退出" + getString(R.string.app_name))
                mLastClickTime = System.currentTimeMillis()
                return true;
            }
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onDestroy() {
        LogUtils.close()
        super.onDestroy()
    }
}